File: BarVelocityTracker/Reps/RepDetector.swift
import Combine
import Foundation

/// State machine that segments the velocity stream into reps.
///
///   idle → eccentric → transition → concentric → topOfRep → eccentric → ...
///
/// A rep is emitted on the `topOfRep` transition. The detector is fed by any
/// `BarMotionSource`, so the same logic works for video-based motion (Phase 1)
/// and BLE/IMU motion (Phase 2).
final class RepDetector: ObservableObject {
    enum Phase: String { case idle, eccentric, transition, concentric, topOfRep }

    @Published private(set) var phase: Phase = .idle
    @Published private(set) var repCount: Int = 0

    /// Emits a `Rep` each time a concentric phase completes.
    let reps = PassthroughSubject<Rep, Never>()

    // Tunable thresholds (see spec)
    var concentricStartThreshold: Double = 0.05    // m/s, upward
    var repEndThreshold: Double = 0.02             // m/s
    var eccentricStartThreshold: Double = -0.05    // m/s, downward
    var minRepGap: TimeInterval = 0.5              // debounce window

    private weak var source: BarMotionSource?
    private var velocityCancellable: AnyCancellable?
    private var positionCancellable: AnyCancellable?

    private var currentPosition: Double = 0
    private var lastVelocity: Double = 0
    private var lastVelocityTime: Date = Date()
    private var lastTimestamp: Date = .distantPast

    // Accumulators for the active concentric phase
    private var concentricStartTime: Date?
    private var concentricStartPosition: Double = 0
    private var concentricVelocities: [Double] = []
    private var concentricPeak: Double = 0
    private var concentricMaxPosition: Double = -.infinity

    init(source: BarMotionSource) {
        self.source = source
        positionCancellable = source.positionPublisher
            .sink { [weak self] in self?.currentPosition = $0 }
        velocityCancellable = source.velocityPublisher
            .sink { [weak self] in self?.ingest(velocity: $0) }
    }

    func reset() {
        phase = .idle
        repCount = 0
        concentricStartTime = nil
        concentricVelocities.removeAll()
        concentricPeak = 0
        concentricMaxPosition = -.infinity
        lastTimestamp = .distantPast
    }

    /// Allow callers to swap detection source after construction (e.g. video → BLE).
    func bind(to source: BarMotionSource) {
        velocityCancellable?.cancel()
        positionCancellable?.cancel()
        self.source = source
        positionCancellable = source.positionPublisher
            .sink { [weak self] in self?.currentPosition = $0 }
        velocityCancellable = source.velocityPublisher
            .sink { [weak self] in self?.ingest(velocity: $0) }
    }

    // MARK: - State machine

    private func ingest(velocity v: Double) {
        let now = Date()
        defer { lastVelocity = v; lastVelocityTime = now }

        switch phase {
        case .idle, .topOfRep:
            if v <= eccentricStartThreshold {
                phase = .eccentric
            } else if v >= concentricStartThreshold {
                // Lifter started concentric from a dead stop (e.g. deadlift).
                beginConcentric(at: now)
            }

        case .eccentric:
            if abs(v) < repEndThreshold {
                phase = .transition
            } else if v >= concentricStartThreshold {
                // Recovery from eccentric straight into concentric (no pause).
                beginConcentric(at: now)
            }

        case .transition:
            if v >= concentricStartThreshold {
                beginConcentric(at: now)
            } else if v <= eccentricStartThreshold {
                phase = .eccentric
            }

        case .concentric:
            concentricVelocities.append(v)
            concentricPeak = max(concentricPeak, v)
            concentricMaxPosition = max(concentricMaxPosition, currentPosition)
            if v < repEndThreshold && concentricPeak > concentricStartThreshold {
                completeRep(at: now)
            }
        }
    }

    private func beginConcentric(at now: Date) {
        // Debounce: ignore if the previous rep finished too recently.
        if now.timeIntervalSince(lastTimestamp) < minRepGap { return }
        phase = .concentric
        concentricStartTime = now
        concentricStartPosition = currentPosition
        concentricVelocities = []
        concentricPeak = 0
        concentricMaxPosition = currentPosition
    }

    private func completeRep(at now: Date) {
        guard let start = concentricStartTime else {
            phase = .topOfRep
            return
        }
        let duration = now.timeIntervalSince(start)
        let mean = concentricVelocities.isEmpty
            ? 0
            : concentricVelocities.reduce(0, +) / Double(concentricVelocities.count)
        let rom = max(0, concentricMaxPosition - concentricStartPosition)

        repCount += 1
        let rep = Rep(index: repCount,
                      meanConcentricVelocity: mean,
                      peakConcentricVelocity: concentricPeak,
                      rangeOfMotion: rom,
                      concentricDuration: duration,
                      timestamp: now)
        reps.send(rep)

        lastTimestamp = now
        concentricStartTime = nil
        phase = .topOfRep
    }
}

File: web/src/lib/repDetector.ts
import type { Rep, RepPhase } from '../types';
import type { KinematicsEngine } from './kinematics';

export interface RepDetectorOptions {
  concentricStartThreshold?: number;
  repEndThreshold?: number;
  eccentricStartThreshold?: number;
  /** Debounce window in seconds. */
  minRepGap?: number;
}

/**
 * State machine that segments the velocity stream into reps. Mirrors the
 * iOS detector. Decoupled from the source: takes a kinematics-shaped
 * subscription so a future BLE/IMU source can drive it instead.
 */
export class RepDetector {
  phase: RepPhase = 'idle';
  repCount = 0;

  private concentricStart: number = 0;
  private concentricStartPosition = 0;
  private concentricVelocities: number[] = [];
  private concentricPeak = 0;
  private concentricMaxPosition = -Infinity;
  private currentPosition = 0;
  private lastEndTimestamp = -Infinity;
  private readonly opts: Required<RepDetectorOptions>;
  private listeners: Set<(rep: Rep) => void> = new Set();
  private unsubscribeVel?: () => void;
  private unsubscribePos?: () => void;

  constructor(source: KinematicsEngine, opts: RepDetectorOptions = {}) {
    this.opts = {
      concentricStartThreshold: opts.concentricStartThreshold ?? 0.05,
      repEndThreshold: opts.repEndThreshold ?? 0.02,
      eccentricStartThreshold: opts.eccentricStartThreshold ?? -0.05,
      minRepGap: opts.minRepGap ?? 0.5,
    };
    this.bind(source);
  }

  bind(source: KinematicsEngine) {
    this.unsubscribeVel?.();
    this.unsubscribePos?.();
    const unsubscribe = source.subscribe({
      onVelocity: (v, t) => this.ingestVelocity(v, t),
      onPosition: (p) => (this.currentPosition = p),
    });
    this.unsubscribeVel = unsubscribe;
  }

  reset() {
    this.phase = 'idle';
    this.repCount = 0;
    this.concentricVelocities = [];
    this.concentricPeak = 0;
    this.concentricMaxPosition = -Infinity;
    this.lastEndTimestamp = -Infinity;
  }

  onRep(listener: (rep: Rep) => void): () => void {
    this.listeners.add(listener);
    return () => {
      this.listeners.delete(listener);
    };
  }

  private ingestVelocity(v: number, t: number) {
    switch (this.phase) {
      case 'idle':
      case 'topOfRep':
        if (v <= this.opts.eccentricStartThreshold) {
          this.phase = 'eccentric';
        } else if (v >= this.opts.concentricStartThreshold) {
          this.beginConcentric(t);
        }
        break;

      case 'eccentric':
        if (Math.abs(v) < this.opts.repEndThreshold) {
          this.phase = 'transition';
        } else if (v >= this.opts.concentricStartThreshold) {
          this.beginConcentric(t);
        }
        break;

      case 'transition':
        if (v >= this.opts.concentricStartThreshold) {
          this.beginConcentric(t);
        } else if (v <= this.opts.eccentricStartThreshold) {
          this.phase = 'eccentric';
        }
        break;

      case 'concentric':
        this.concentricVelocities.push(v);
        if (v > this.concentricPeak) this.concentricPeak = v;
        if (this.currentPosition > this.concentricMaxPosition) {
          this.concentricMaxPosition = this.currentPosition;
        }
        if (
          v < this.opts.repEndThreshold &&
          this.concentricPeak > this.opts.concentricStartThreshold
        ) {
          this.completeRep(t);
        }
        break;
    }
  }

  private beginConcentric(t: number) {
    if (t - this.lastEndTimestamp < this.opts.minRepGap) return;
    this.phase = 'concentric';
    this.concentricStart = t;
    this.concentricStartPosition = this.currentPosition;
    this.concentricVelocities = [];
    this.concentricPeak = 0;
    this.concentricMaxPosition = this.currentPosition;
  }

  private completeRep(t: number) {
    const duration = t - this.concentricStart;
    const mean =
      this.concentricVelocities.length === 0
        ? 0
        : this.concentricVelocities.reduce((a, b) => a + b, 0) /
          this.concentricVelocities.length;
    const rom = Math.max(0, this.concentricMaxPosition - this.concentricStartPosition);

    this.repCount += 1;
    const rep: Rep = {
      id: crypto.randomUUID(),
      index: this.repCount,
      meanConcentricVelocity: mean,
      peakConcentricVelocity: this.concentricPeak,
      rangeOfMotion: rom,
      concentricDuration: duration,
      timestamp: Date.now(),
      videoStart: this.concentricStart,
      videoEnd: t,
    };
    for (const l of this.listeners) l(rep);

    this.lastEndTimestamp = t;
    this.phase = 'topOfRep';
  }
}
